### PR TITLE
[SPARK-28565][SQL] DataFrameWriter saveAsTable support for V2 catalogs

### PR DESCRIPTION
**Pull request by brkyvz**
## What changes were proposed in this pull request?

Adds support for V2 catalogs and the V2SessionCatalog for V2 tables for saveAsTable.
If the table can resolve through the V2SessionCatalog, we use SaveMode for datasource v1 for backwards compatibility to select the code path we're going to hit.

Depending on the SaveMode:
 - SaveMode.Append:
     a) If table exists: Use AppendData.byName
     b) If table doesn't exist, use CTAS (ignoreIfExists = false)
 - SaveMode.Overwrite: Use RTAS (orCreate = true)
 - SaveMode.Ignore: Use CTAS (ignoreIfExists = true)
 - SaveMode.ErrorIfExists: Use CTAS (ignoreIfExists = false)

## How was this patch tested?

Unit tests in DataSourceV2DataFrameSuite

### CODE REVIEW

**Comment by brkyvz** [comment]
cc @jzhuge @rdblue @cloud-fan

**Comment by brkyvz** [comment]
cc @jose-torres as well

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2DataFrameSuite.scala [hunk 1]
```diff
@@ -104,4 +137,169 @@ class DataSourceV2DataFrameSuite extends QueryTest with SharedSQLContext with Be
...
+  testQuietly("saveAsTable: with defined catalog + ignore mode and table exists") {
+    val t1 = "testcat.ns1.ns2.tbl"
+    withTable(t1) {
+      sql(s"CREATE TABLE $t1 USING foo AS SELECT 'c', 'd'")
+      spark.table("source").write.mode("ignore").saveAsTable(t1)
+      checkAnswer(spark.table(t1), Seq(Row("c", "d")))
+    }
+  }
+
+  sessionCatalogTest("saveAsTable and v2 table - table doesn't exist") { session =>
```
Comment (jzhuge) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2DataFrameSuite.scala]: Consider moving all session catalog tests to a new suite file?
This way, the entire suite can have a Spark session that has v2 session catalog defined.
We call also add new insertInto test cases for session catalog there.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -648,8 +648,13 @@ class Analyzer(
           if catalog.isTemporaryTable(ident) =>
         u // temporary views take precedence over catalog table names
 
-      case u @ UnresolvedRelation(CatalogObjectIdentifier(Some(catalogPlugin), ident)) =>
-        loadTable(catalogPlugin, ident).map(DataSourceV2Relation.create).getOrElse(u)
+      case u @ UnresolvedRelation(CatalogObjectIdentifier(maybeCatalog, ident)) =>
+        // First try loading the table with a loadable catalog, then fallback to the session
+        // catalog if that exists
+        maybeCatalog.flatMap(loadTable(_, ident))
```
Comment (jzhuge) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: How about?
```
        maybeCatalog.orElse(sessionCatalog)
          .flatMap(loadTable(_, ident))
```

Comment (brkyvz) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: That's the tricky thing. We need to see if we can load using the defined catalog first. If the catalog is defined, but it doesn't return a table, then we miss out on using the session catalog.

Comment (jzhuge) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: When a user tries to use table `cat1.ns1.ns2.tbl`, where
1. catalog `cat1` exists, but without table `ns1.ns2.tbl`
2. v2 session catalog is specified and it has a table named `cat1.ns1.ns2.tbl`

IMHO, Spark should throw no table found exception. This is consistent with the follow statement in #24768:

> A session catalog should be used when the v1 catalog is responsible for tables with no catalog in the table identifier.

 Maybe I missed some discussions about fallback?

**Comment by cloud-fan** [comment]
I'd like to call a discussion about the v2 session catalog. I think v2 session catalog is an extension of the builtin catalog(hive catalog). It allows users to add some hooks when Spark sends metadata requests to the builtin catalog.

With this in mind, the table lookup logic should be:
1. if the table identifier specifies the catalog name, look up the table from that catalog, and fail if table not found
2. if the  table identifier do not have catalog name, then look up the table from default catalog
    - if the default catalog has been set(i.e. it's a custom catalog), look up table from it and fail if table not found
    - if the default catalog has not been set(i.e. it's the builtin catalog), look up table from v2 session catalog to apply the hooks

Similar logic should be applied to table creation as well.

I think this PR makes sense: if we need to go to v2 session catalog for table creation, we should do the same for table lookup as well. But I do agree with @jzhuge about https://github.com/apache/spark/pull/25330#discussion_r309950145

what do you think? @rdblue @mccheah

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 2 above)
Comment (rdblue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: > That's the tricky thing. We need to see if we can load using the defined catalog first. If the catalog is defined, but it doesn't return a table, then we miss out on using the session catalog.

I disagree.

Determining which catalog is responsible for an identifier is independent of the catalog results. If `maybeCatalog` is `None`, then the session catalog is responsible for the identifier. Whether to use v1 or v2 after that point depends on the provider.

**Comment by rdblue** [comment]
@cloud-fan, I think we are saying the same thing. The identifier supplied by the user determines the catalog that should be used (considering session catalog and v2 session catalog to be the same catalog). Then v2 or v1 session catalog is used depending on what the provider implementation needs.

We already do this for `CREATE TABLE` SQL: https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala#L51-L72

That first rule only matches if there is no v2 catalog (not set in the identifier and no default), and if the provider is a v1 provider. The second rule matches all cases where v2 will be used. `CatalogObjectIdentifier` returns a catalog if there is an explicit catalog or a default. If that is `None`, then the session catalog is responsible for the identifier and because the first rule didn't match, we know that the provider is v2, so the v2 session catalog is used.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 2 above)
Comment (brkyvz) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: @rdblue Imagine the following use case. Let's say as the data science team, I have a database on the Hive MetaStore called `prod`. I have tables named like: `prod.tbl1`, `prod.tbl2`. Now, some other team (maybe data engineering team), creates a v2 catalog called `prod` and adds it to my environment.

Won't the `prod` catalog hijack all my requests, and start failing to resolve the tables I had declared before, which I was expecting the V2SessionCatalog to resolve? I, as an unaware user, suddenly I have all my jobs failing and chaos ensues.

Comment (rdblue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Yes. A conflict could break queries in both directions when there is a conflict between namespace `prod` and catalog `prod`:
* If a namespace takes precedence and a new one is created, it will break queries that use the catalog
* If a catalog takes precedence and a new one is created, it will break queries that use the namespace

When we wrote the SPIP, the choice was to make catalog take precedence because:
1. Any user can create a namespace globally in the catalog, possibly breaking other users
2. Users can't globally create catalogs -- that's done by administrators -- so the impact is limited to their own jobs where they can add Spark configs
3. Global catalogs are created by administrators and this happens more rarely

In short, we expect fewer problems when catalogs take precedence.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 3]
```diff
@@ -485,7 +486,80 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
...
+    }.getOrElse(Seq.empty[Transform])
+    val partitionTransforms = partitioning ++ bucketing
+
+    val tableOpt = try Option(catalog.loadTable(ident)) catch {
+      case _: NoSuchTableException => None
+    }
+
+    val command = (mode, tableOpt) match {
+      case (SaveMode.Append, Some(table)) =>
+        AppendData.byName(DataSourceV2Relation.create(table), df.logicalPlan)
```
Comment (jzhuge) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Add unit test to verify byName not byPosition

**Comment by brkyvz** [comment]
Thanks for the review @jzhuge and @rdblue . Addressed your comments.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 4]
```diff
@@ -648,8 +648,11 @@ class Analyzer(
           if catalog.isTemporaryTable(ident) =>
         u // temporary views take precedence over catalog table names
 
-      case u @ UnresolvedRelation(CatalogObjectIdentifier(Some(catalogPlugin), ident)) =>
-        loadTable(catalogPlugin, ident).map(DataSourceV2Relation.create).getOrElse(u)
+      case u @ UnresolvedRelation(CatalogObjectIdentifier(maybeCatalog, ident)) =>
+        maybeCatalog.orElse(sessionCatalog)
```
Comment (jzhuge) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: This changes the SELECT path, so more unit tests for table relation are needed, e.g.:
```
  test("Relation: basic") {
    val t1 = "tbl"
    withTable(t1) {
      sql(s"CREATE TABLE $t1 USING $v2Format AS SELECT id, data FROM source")
      checkAnswer(spark.table(t1), spark.table("source"))
      checkAnswer(sql(s"TABLE $t1"), spark.table("source"))
      checkAnswer(sql(s"SELECT * FROM $t1"), spark.table("source"))
    }
  }
```

IMHO this change v2 session catalog and related tests in `DataSourceV2SessionDataFrameSuite` can be split to a separate PR with additional unit tests for SQL SELECT and spark.table().

Another PR to update `ResolveInsertInto` to support v2 session catalog in SQL INSERT INTO and DFW.insertInto.

**Review by jzhuge — CHANGES_REQUESTED** [review]
Mostly LGTM. However, v2 session catalog changes feel like a separate PR.

**Comment by brkyvz** [comment]
Sounds good. I can separate the session catalog things. That's a valid point around requiring more tests around reads.

**Comment by brkyvz** [comment]
Or would you mind if those are added as a follow up in supporting INSERT INTO as well?

**Comment by jzhuge** [comment]
It'd better to group READ/SELECT and saveAsTable together because they both are affected by the change in `ResolveTables`; while INSERT is affected by the change in `ResolveInsertInto`.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 5]
```diff
@@ -485,7 +486,80 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
...
+          partitionTransforms,
+          df.queryExecution.analyzed,
+          Map.empty,            // properties can't be specified through this API
+          extraOptions.toMap,
+          orCreate = true)      // Create the table if it doesn't exist
+
+      case (other, _) =>
+        // We have a potential race condition here in AppendMode, if the table suddenly gets
+        // created between our existence check and physical execution, but this can't be helped
+        // in any case.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: If we treat `asAsTable` as one atomic operation, I think AppendMode means "create or append table". Since "create or append table" is not a standard SQL commannd, maybe it's fine to ignore this race condition.

**Comment by brkyvz** [comment]
cc @jzhuge Addressed your comments.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 6]
```diff
@@ -485,7 +485,65 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
    * @since 1.4.0
    */
   def saveAsTable(tableName: String): Unit = {
-    saveAsTable(df.sparkSession.sessionState.sqlParser.parseTableIdentifier(tableName))
+    import df.sparkSession.sessionState.analyzer.{AsTableIdentifier, CatalogObjectIdentifier}
+    import org.apache.spark.sql.catalog.v2.CatalogV2Implicits._
+
+    val session = df.sparkSession
```
Comment (jzhuge) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Nit: the local variable is used only once.

**Review by jzhuge — APPROVED** [review]
LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 7]
```diff
@@ -485,7 +485,65 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
...
+
+    session.sessionState.sqlParser.parseMultipartIdentifier(tableName) match {
+      case CatalogObjectIdentifier(Some(catalog), ident) =>
+        saveAsTable(catalog.asTableCatalog, ident, modeForDSV2)
+
+      case AsTableIdentifier(tableIdentifier) =>
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: This should check whether the provider is v2 and use v2 and the V2SessionCatalog if it is.

```scala
case CatalogObjectIdentifier(None, ident) if isV1(provider) =>
  saveAsTable(v2SessionCatalog, ident, modeForDSV2)
```

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 8]
```diff
@@ -485,7 +485,65 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
...
+        saveAsTable(tableIdentifier)
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: What about cases where `AsTableIdentifier` won't match? For example, `saveAsTable("ns1.ns2.table")`. For those cases, we need a `case _ =>` that throws an exception because the table name is not supported by the session catalog.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 9]
```diff
@@ -485,7 +485,65 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
...
+
+    val tableOpt = try Option(catalog.loadTable(ident)) catch {
+      case _: NoSuchTableException => None
+    }
+
+    val command = (mode, tableOpt) match {
+      case (SaveMode.Append, Some(table)) =>
+        AppendData.byName(DataSourceV2Relation.create(table), df.logicalPlan)
+
+      case (SaveMode.Overwrite, _) =>
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: I think that the behavior here is to truncate and write, not replace.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 7 above)
Comment (brkyvz) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: we separated the session catalog to a separate PR to get this in and add more tests to the session catalog support

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 9 above)
Comment (brkyvz) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: we may want to change the table properties though, such as partitioning and schema, wouldn't we?

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 6 above)
Comment (brkyvz) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: I'll use it in the follow up :)

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 7 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Sounds good to me.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 9 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: We need to match the behavior of v1 file sources. Do we know what that behavior is?

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 7 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Sounds fine to me, but let's file a JIRA and mark it as a blocker for 3.0.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 9 above)
Comment (brkyvz) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: As discussed in DSV2 sync, the old behavior was to drop the old table and create a new one, therefore the Replace behavior here works.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 7 above)
Comment (brkyvz) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Created SPARK-28666

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 10]
```diff
@@ -485,7 +490,71 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
...
+      case CatalogObjectIdentifier(Some(catalog), ident) =>
+        saveAsTable(catalog.asTableCatalog, ident, modeForDSV2)
+
+      case AsTableIdentifier(tableIdentifier) =>
+        saveAsTable(tableIdentifier)
+
+      case other =>
+        // TODO(SPARK-28666): This should go through V2SessionCatalog
+        throw new UnsupportedOperationException(
+          s"Couldn't find a catalog to handle the identifier ${other.quoted}.")
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: I think this is an analysis error. The catalog was None, so it belongs to the session catalog. But the session catalog doesn't support namespaces with more than one part, so it is an invalid identifier for that catalog.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 7 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Could you add a TODO comment to this as well?

**Comment by rdblue** [comment]
I had a couple of minor comments about documenting what will be done when this supports the v2 session catalog and throwing an exception. Once those are fixed, +1.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 10 above)
Comment (brkyvz) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: ah gotcha, that's the 4th case... I'd like to make that change along with the V2SessionCatalog support. Right now we just couldn't find a catalog for it, so... (Maybe they forgot to specify a default catalog).

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 11]
```diff
@@ -374,8 +375,12 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
     df.sparkSession.sessionState.sqlParser.parseMultipartIdentifier(tableName) match {
       case CatalogObjectIdentifier(Some(catalog), ident) =>
         insertInto(catalog, ident)
+      // TODO(SPARK-28667): Support the V2SessionCatalog
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Not related to this PR, but just a note. I've been working on V2SessionCatalog improvement for a while, and one issue I found is: it's not easy to implement "if the table provider is v1, go to v1 session catalog. if the table provider is v2, go to v2 session catalog".

It's easy to do it for CREATE TABLE, because the table provider is known at the beginning. But it's hard for SELECT and INSERT, as we need to look up the table from Hive catalog to get the table provider.

I'd expect to have 2 analyzer rules to do it:
1. one rule in sql/catalyst, which resolves `UnresolvedRelation` to `UnresolvedCatalogRelation` by looking up the table from hive catalog
2. one rule in sql/core, which resolves `UnresolvedCatalogRelation` to `DataSourceV2Relation`, if the table provider is v2. This has to be in sql/core because `DataSource.lookupDataSource` is in sql/core. We have a rule `FindDataSourceTable` that resolves `UnresolvedCatalogRelation` to v1 relation if the table provider is v1.

I think `DataFrameWriter` should produce `...Statement` plans, to reuse the analyzer rules for all the things.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2DataFrameSuite.scala [hunk 12]
```diff
@@ -141,4 +141,66 @@ class DataSourceV2DataFrameSuite extends QueryTest with SharedSQLContext with Be
       }
     }
   }
+
+  testQuietly("saveAsTable: table doesn't exist => create table") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2DataFrameSuite.scala]: just for curiosity, why do we need `testQuietly` here?

**Review by cloud-fan — APPROVED** [review]
LGTM except a few minor comments

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 10 above)
Comment (brkyvz) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: That's true, but that could also mean that they forgot to configure their default catalog too, right?
I can throw a better error when the SessionCatalog code path is implemented, since we need that before we can even say that `ns1.ns2.table` cannot be handled by the session catalog.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 11 above)
Comment (brkyvz) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Well, the V2SessionCatalog already produces a `CatalogTableAsV2`, which gets unwrapped to the `UnresolvedCatalogRelation` already in `DataSourceResolution`, and then it can get resolved to the V1 code paths. I've also had a similar problem when trying to implement V2SessionCatalog support for `ALTER TABLE`. In that case I decided that it should reject and fallback to V1 plans if the loaded table is a `CatalogTableAsV2`.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2DataFrameSuite.scala (hunk 12 above)
Comment (brkyvz) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2DataFrameSuite.scala]: It was before merging Gengliang's "Catalogs.load should work for inbuilt catalogs" PR. It may not be needed now, but I don't think it's a big deal

**Comment by brkyvz** [comment]
Thanks all for the reviews! Merging to master :)

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 11 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: > V2SessionCatalog already produces a CatalogTableAsV2

Yes, for both v1 and v2 provider. So we still need a rule to check the table provider, and either unwrap it to a `UnresolvedCatalogRelation`, or create the actual v2 table.

Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: > Yes, for both v1 and v2 provider. So we still need a rule to check the table provider, and either unwrap it to a UnresolvedCatalogRelation, or create the actual v2 table.

This should be done in `loadTable`. The original implementation called `TableProvider` to create the table, but this path is currently broken. I think that `TableProvider` needs to be improved so that it works instead of adding more rules to convert from `CatalogTableAsV2`. 

The catalog should return the correct `Table` instance for all v2 tables. Spark shouldn't convert between v2 table instances.
